pkg/controller/master/addon/common.go

```diff
@@ -0,0 +1,402 @@
+	return h.addon.UpdateStatus(a)
+}
+
+func (h *Handler) updateAddonDisableStatus(aObj *harvesterv1.Addon, addonStatus harvesterv1.AddonState, status corev1.ConditionStatus, reason, message string) (*harvesterv1.Addon, error) {
+	a := aObj.DeepCopy()
+	a.Status.Status = addonStatus
+	h.updateAddonConditions(a, harvesterv1.AddonDisableCondition, status, reason, message, addonStatus == harvesterv1.AddonDisabling)
+	// clear failed condition
+	if addonStatus == harvesterv1.AddonDisabling {
+		h.updateAddonConditions(a, harvesterv1.AddonFailedCondition, status, reason, message, true)
+	}
+	// log error message
+	if message != "" {
+		h.updateAddonConditions(a, harvesterv1.AddonFailedCondition, status, reason, message, false)
+	}
+	logrus.Debugf("addon %s disable operation will be set to new status %s", a.Name, addonStatus)
+	return h.addon.UpdateStatus(a)
+}
+
+func (h *Handler) updateAddonUpdateStatus(aObj *harvesterv1.Addon, addonStatus harvesterv1.AddonState, status corev1.ConditionStatus, reason, message string) (*harvesterv1.Addon, error) {
+	a := aObj.DeepCopy()
+	a.Status.Status = addonStatus
+	h.updateAddonConditions(a, harvesterv1.AddonUpdateCondition, status, reason, message, addonStatus == harvesterv1.AddonUpdating)
+	// clear failed condition
+	if addonStatus == harvesterv1.AddonUpdating {
+		h.updateAddonConditions(a, harvesterv1.AddonFailedCondition, status, reason, message, true)
+	}
+	// log error message
+	if message != "" {
+		h.updateAddonConditions(a, harvesterv1.AddonFailedCondition, status, reason, message, false)
+	}
+	logrus.Debugf("addon %s update operation will be set to new status %s", a.Name, addonStatus)
+	return h.addon.UpdateStatus(a)
+}
+
+// directly update the status to final, no state transfer
+func (h *Handler) updateAddonUpdateStatusDirectlyToFinal(aObj *harvesterv1.Addon, addonStatus harvesterv1.AddonState, status corev1.ConditionStatus, reason, message string) (*harvesterv1.Addon, error) {
+	a := aObj.DeepCopy()
+	a.Status.Status = addonStatus
+	// in such case, set the LastTransitionTime as well
+	h.updateAddonConditions(a, harvesterv1.AddonUpdateCondition, status, reason, message, true)
+	// log error message, normally, here it has no error message
+	if message != "" {
+		h.updateAddonConditions(a, harvesterv1.AddonFailedCondition, status, reason, message, true)
+	}
+	logrus.Debugf("addon %s update operation will be set to final status directly %s", a.Name, a.Status.Status)
+	return h.addon.UpdateStatus(a)
+}
+
+func (h *Handler) setAddonConditionLastTransitionTime(aObj *harvesterv1.Addon, cond condition.Cond) {
+	for i := range aObj.Status.Conditions {
+		if aObj.Status.Conditions[i].Type == cond {
+			aObj.Status.Conditions[i].LastUpdateTime = metav1.Now().UTC().Format(time.RFC3339)
+			return
+		}
+	}
+
+	// not find
+	if len(aObj.Status.Conditions) == 0 {
+		aObj.Status.Conditions = make([]harvesterv1.Condition, 1)
```

Comment:
Could we use init condition instead of the empty condition? The empty condition seems like a bug.

Reply:
could you explain a bit more? what do you mean `init` condition

Reply:
`Addon.status.conditions` will appear as including an empty condition when enable for the first time, it should add a new type instead the empty conditions, e.g.: Initialized, as follow:
```yaml
status:
  conditions:
    # Empty condition
    - status: ''
      type: ''
    # Initialized condition
    - lastUpdateTime: '2023-07-06T06:25:23Z'
      status: 'True'
      type: 'Initialized'
    - lastUpdateTime: '2023-07-06T06:25:23Z'
      status: ''
      type: Enable
  status: AddonEnabled
```